### PR TITLE
Demonstrate use of new `LockTime` type to replace u32 in `After`.

### PR DESCRIPTION
**Pull request by tcharding**
This PR shows how we might use the new `LockTime` type from `rust-bitcoin`. (https://github.com/rust-bitcoin/rust-bitcoin/pull/994)

This PR is not intended to be a merge candidate.

Please note, this PR has changed a lot, comments below are mostly stale. 

Currently all test pass when run locally, the bitcoin dependency has to be set in _all_ the other dependencies (bitcoind, rust-bitcoincore-rpc). I only did this locally - you'll just have to believe me :)

### CODE REVIEW

**Comment by apoelstra** [comment]
Nice! concept ACK.

**Comment by tcharding** [comment]
Patch 9 could actually go in rust-bitcoin, there is nothing specific to miniscript in it?

**Comment by apoelstra** [comment]
Oh, yeah, nice idea. I'd be in favor of bringing timelocks into rust-bitcoin. Not sure how the other maintainers would feel.

Then we could add accessor functions to `Transaction` that can read the sequence numbers and locktime..

**Comment by tcharding** [comment]
> Then we could add accessor functions to Transaction that can read the sequence numbers and locktime..

Please review patch 2 of https://github.com/rust-bitcoin/rust-bitcoin/pull/994, is this what you meant?

**Comment by apoelstra** [comment]
This looks great! I did not review it in detail.

File: src/miniscript/types/mod.rs [hunk 1]
```diff
@@ -820,7 +822,7 @@ impl Property for Type {
                 // Note that for CLTV this is a limitation not of Bitcoin but Miniscript. The
                 // number on the stack would be a 5 bytes signed integer but Miniscript's B type
                 // only consumes 4 bytes from the stack.
-                if t == 0 || (t & SEQUENCE_LOCKTIME_DISABLE_FLAG) != 0 {
+                if t.to_consensus_u32() == 0 || (t.to_consensus_u32() & SEQUENCE_LOCKTIME_DISABLE_FLAG) != 0 {
```
Comment (Kixunil) [inline src/miniscript/types/mod.rs:825]: Perhaps a reason to have `is_zero()` method in `LockTime`? Anyway, this looks nicer to me: `t == LockTime::ZERO`

Also WTF is sequence mixed with lock time here?! Looks very wrong to me!

File: src/miniscript/astelem.rs [hunk 2]
```diff
@@ -622,7 +623,7 @@ impl<Pk: MiniscriptKey, Ctx: ScriptContext> Terminal<Pk, Ctx> {
                 .push_slice(&Pk::hash_to_hash160(hash)[..])
                 .push_opcode(opcodes::all::OP_EQUALVERIFY),
             Terminal::After(t) => builder
-                .push_int(t as i64)
+                .push_int(t.to_consensus_u32() as i64)
```
[the remark this answers is not in the document]
Comment (dpc) [reply src/miniscript/astelem.rs:626]: What's the point of having long-ass `to_consensus_u32` and then being able to just `.into()` :D

File: src/miniscript/types/mod.rs (hunk 1 above)
Comment (dpc) [reply src/miniscript/types/mod.rs:825]: > Anyway, this looks nicer to me: `t == LockTime::ZERO`

:+1:

File: src/miniscript/astelem.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply src/miniscript/astelem.rs:626]: `i64::from()` is also a possibility. My point is to avoid `as`.

[the remark this answers is not in the document]
Comment (dpc) [reply src/miniscript/astelem.rs:626]: My point is maybe we should we should get rid of that `Into`/`From` if we don't want users to willy-nilly convert to raw integer types.

[the remark this answers is not in the document]
Comment (tcharding) [reply src/miniscript/astelem.rs:626]: I removed the `From<u32>` impls a few iterations ago for exactly the reason you state @dpc. You off the cuff suggestion @Kixunil makes me think we would have been fine to use `to_u32` all along and not had the extremely long conversation about `to_u32` and `to_consensus_u32` :) I guess naming this _is_ hard.

File: src/miniscript/astelem.rs [hunk 3]
```diff
@@ -757,7 +758,7 @@ impl<Pk: MiniscriptKey, Ctx: ScriptContext> Terminal<Pk, Ctx> {
         match *self {
             Terminal::PkK(ref pk) => Ctx::pk_len(pk),
             Terminal::PkH(..) | Terminal::RawPkH(..) => 24,
-            Terminal::After(n) => script_num_size(n as usize) + 1,
+            Terminal::After(n) => script_num_size(n.to_consensus_u32() as usize) + 1,
```
[the remark this answers is not in the document]
Comment (tcharding) [reply src/miniscript/astelem.rs:761]: https://github.com/rust-bitcoin/rust-miniscript/pull/442

File: src/miniscript/types/mod.rs (hunk 1 above)
Comment (tcharding) [reply src/miniscript/types/mod.rs:825]: Sure thing.

File: src/policy/concrete.rs [hunk 4]
```diff
@@ -608,7 +620,16 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
                     Ok(())
                 }
             }
-            Policy::After(n) | Policy::Older(n) => {
+            Policy::After(n)  => {
+                if n.to_consensus_u32() == 0 {
+                    Err(PolicyError::ZeroTime)
+                } else if n.to_consensus_u32() > 2u32.pow(31) {
```
[the remark this answers is not in the document]
Comment (tcharding) [reply src/policy/concrete.rs:626]: Correct me if I'm wrong @sanket1729 or @apoelstra but I think this is a miniscript restriction not a locktime restriction. If that's the case I don't think it should be part of `rust-bitcoin`.

File: src/policy/semantic.rs [hunk 5]
```diff
@@ -634,12 +647,21 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
...
         match self {
             Policy::Threshold(k, subs) => {
                 let mut new_subs: Vec<_> = subs.into_iter().map(Policy::sorted).collect();
-                new_subs.sort();
+                new_subs.sort_by(|a, b| {
+                    // partial_cmp on the enum should give an ordering for any a/b that are
+                    // different as well as for any a/b that are the same except After/After.
+                    match (a, b) {
+                        (After(a), After(b)) => a.to_consensus_u32().cmp(&b.to_consensus_u32()),
+                        (a, b) => a.partial_cmp(b).unwrap_or(Ordering::Equal),
```
[the remark this answers is not in the document]
Comment (tcharding) [reply src/policy/semantic.rs:660]: Yes I thought the same originally but I don't believe we can do so, e.g.
```rust
(Trivial, Unsatisfiable) => a.cmp(&b),
```
`cmp` is not implemented for `Policy` (that's the whole thing we are trying to get around). Or am I missing something?

**Comment by tcharding** [comment]
Thanks for the review @Kixunil, I believe you may now start to see some of my pain in doing this work :) Its not trivial to code up the `LockTime` API using miniscript code _without_ knowing miniscript well (which I do not).

File: src/miniscript/astelem.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply src/miniscript/astelem.rs:624]: We're talking completely different types. Conversion from `u32` to `i64` is provided by `core` and it's a completely obvious conversion preserving the value as the same number.

File: src/miniscript/types/mod.rs (hunk 1 above)
Comment (Kixunil) [reply src/miniscript/types/mod.rs]: > ecause associated consts don't exist for enums

Did you miss my comment before demonstrating it's not the case?

File: src/policy/concrete.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply src/policy/concrete.rs:632]: Then it should be a constant in `miniscript` :)

File: src/policy/semantic.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply src/policy/semantic.rs:666]: We can do the exact thing derive does: return `Ordering::Greater`.

File: src/policy/concrete.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/policy/concrete.rs:632]: You are correct @tcharding.

I would suggest putting the constant in rust-bitcoin as `MAX_SCRIPTNUM_VALUE` or something.

File: src/miniscript/astelem.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply src/miniscript/astelem.rs:624]: Will add, thanks.

File: src/miniscript/types/mod.rs (hunk 1 above)
Comment (tcharding) [reply src/miniscript/types/mod.rs]: Now includes `LockTime::ZERO` :)

File: src/policy/concrete.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply src/policy/concrete.rs:632]: Added to `rust-bitcoin/src/blockdata/constants.rs`.

**Comment by tcharding** [comment]
`LockTime` has merged already, closing.

**Pull request closed without merging**